pulsar-client-kafka-compat/src/main/java/org/apache/kafka/clients/consumer/PulsarKafkaConsumer.java

```diff
@@ -0,0 +1,482 @@
+            String topicName = DestinationName.get(topicPartition.topic()).getPartition(topicPartition.partition())
+                    .toString();
+            org.apache.pulsar.client.api.Consumer consumer = consumers.get(topicName);
+
+            MessageId messageId = getMessageId(offsetAndMetadata.offset());
+            futures.add(consumer.acknowledgeCumulativeAsync(messageId));
+        });
+
+        futures.forEach(CompletableFuture::join);
+    }
+
+    @Override
+    public void commitAsync() {
+        lastMessageId.forEach((consumer, messageId) -> {
+            consumer.acknowledgeCumulativeAsync(messageId);
+        });
+    }
+
+    @Override
+    public void commitAsync(OffsetCommitCallback callback) {
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
+        lastMessageId.forEach((consumer, messageId) -> {
+            DestinationName dn = DestinationName.get(consumer.getTopic());
+            int partition = dn.isPartitioned() ? dn.getPartitionIndex() : 0;
+
+            TopicPartition tp = new TopicPartition(dn.getPartitionedTopicName(), partition);
+            OffsetAndMetadata om = new OffsetAndMetadata(getOffset(messageId));
+            offsets.put(tp, om);
+
+            futures.add(consumer.acknowledgeCumulativeAsync(messageId));
+        });
+
+        FutureUtil.waitForAll(futures).handle((v, throwable) -> {
+            callback.onComplete(offsets, throwable != null ? new Exception(throwable) : null);
+            return null;
+        });
+    }
+
+    @Override
+    public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        List<CompletableFuture<Void>> futures = new ArrayList<>();
+
+        offsets.forEach((topicPartition, offsetAndMetadata) -> {
+            String topicName = DestinationName.get(topicPartition.topic()).getPartition(topicPartition.partition())
+                    .toString();
+            org.apache.pulsar.client.api.Consumer consumer = consumers.get(topicName);
+
+            futures.add(consumer.acknowledgeCumulativeAsync(getMessageId(offsetAndMetadata.offset())));
+        });
+
+        FutureUtil.waitForAll(futures).handle((v, throwable) -> {
+            callback.onComplete(offsets, throwable != null ? new Exception(throwable) : null);
+            return null;
+        });
+    }
+
+    @Override
+    public void seek(TopicPartition partition, long offset) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void seekToBeginning(Collection<TopicPartition> partitions) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void seekToEnd(Collection<TopicPartition> partitions) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long position(TopicPartition partition) {
+        throw new UnsupportedOperationException();
```

Comment:
I think this can be implemented, no? you kept the lastMessageId?

Reply:
Yes, it can be implemented, also the `committed()` method as well. I just need to change the map, since the key is currently a `Consumer` instance.